torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -66,37 +67,25 @@ std::vector<at::Device> getDeviceList(const std::vector<at::Tensor>& tensors) {
   return res;
 }
 
-// Helper that lets the input ncclStreams to wait for the THC stream
+// Helper that lets the input ncclStreams to wait for the current stream
 void syncStreams(
-    THCState* thcState,
     const std::vector<at::Device>& devices,
-    std::vector<CUDAEvent>& ncclEvents,
+    std::vector<at::cuda::CUDAEvent>& ncclEvents,
     std::vector<at::cuda::CUDAStream>& ncclStreams) {
-  at::cuda::CUDAGuard gpuGuard;
   for (size_t i = 0; i < devices.size(); ++i) {
-    gpuGuard.set_index(devices[i].index());
-    auto currentThcStream =
-        THCState_getCurrentStreamOnDevice(thcState, devices[i].index());
     at::cuda::CUDAStream& ncclStream = ncclStreams[i];
-    CUDAEvent& ncclEvent = ncclEvents[i];
-
-    C10D_CUDA_CHECK(cudaEventRecord(ncclEvent.getEvent(), currentThcStream));
-    C10D_CUDA_CHECK(
-        cudaStreamWaitEvent(ncclStream.stream(), ncclEvent.getEvent(), 0));
+    at::cuda::CUDAEvent& ncclEvent = ncclEvents[i];
+    ncclEvent.record(at::cuda::getCurrentCUDAStream(devices[i].index()));
+    ncclEvent.block(ncclStream);
```

Comment:
This is less optimal than the previous case (related to the device guard in CUDAEvent::record). Instead of updating a single guard you end up setting/unsetting the device N times. This is `devices.size()` times more `cudaSetDevice` calls.

Reply:
You are correct. Let me think about a way to fix this. One simple approach may be to expand the at::cuda::CUDAEvent API to accept an existing DeviceGuard.

Reply:
Two options:
- Create versions of record() and block() that accept an at::cuda::CUDAGuard. The guard is set to the appropriate device. This may be a little strange since currently the device is reset after these calls. That is, if you're on device X and the stream is on device Y, you can block the stream and after that call you're back on device X. In this plan you'd still be on device Y. 
- Create versions of record() and block() that accept a bool of whether they should set the device or not. Then you can guard around these calls yourself if desired. The bool should also make the necessity of guarding clear (default would be to use guards in CUDAEvent). 

Would either of these be acceptable? I would like to suggest the latter. Minimal code change and the UX seems clear.

Reply:
@pietern Is `syncStreams` being called a lot? If it's not, I'm inclined to just eat the cost. In any case, latter seems preferred.